src/webui/api/torrentscontroller.cpp

```diff
@@ -1125,22 +1125,44 @@ void TorrentsController::addTrackersAction()
 
 void TorrentsController::editTrackerAction()
 {
-    requireParams({u"hash"_s, u"origUrl"_s, u"newUrl"_s});
+    requireParams({u"hash"_s, u"origUrl"_s});
 
     const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
     const QString origUrl = params()[u"origUrl"_s];
-    const QString newUrl = params()[u"newUrl"_s];
+    const std::optional<QString> newUrlParam = getOptionalString(params(), u"newUrl"_s);
+    const std::optional<QString> newTierParam = getOptionalString(params(), u"tier"_s);
 
     BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
     if (!torrent)
         throw APIError(APIErrorType::NotFound);
+    if (!newUrlParam && !newTierParam)
+        throw APIError(APIErrorType::BadParams, u"Must specify at least one of [newUrl, tier]"_s);
+
+    std::optional<int> newTier;
+    if (newTierParam)
+    {
+        bool ok = false;
+        const auto number = newTierParam.value().toInt(&ok);
+        if (!ok)
+            throw APIError(APIErrorType::BadParams, u"New tier must be an integer"_s);
+        if ((number < 0) || (number > 255))
+            throw APIError(APIErrorType::BadParams, u"New tier must be between 0 and 255"_s);
+        newTier = number;
+    }
 
     const QUrl origTrackerUrl {origUrl};
-    const QUrl newTrackerUrl {newUrl};
-    if (origTrackerUrl == newTrackerUrl)
+    std::optional<QUrl> newTrackerUrl;
+    if (newUrlParam)
+    {
+        const QUrl url = newUrlParam.value();
+        if (!url.isValid())
+            throw APIError(APIErrorType::BadParams, u"New tracker URL is invalid"_s);
+        if (url != origTrackerUrl)
```

Comment:
Actually we can consider the case of `newUrl == url` as a "bad params", no?

Reply:
Potentially yes though when using this form in the WebUI, it seems strange that if you open the Edit Tracker dialog, and then click Save without changing anything, it displays an error. I think it makes sense to just consider that a no-op, which would mean the API returning a 200/204.

Reply:
I think it's more common for an "edit" endpoint to treat a lack of changes as a no-op, rather than an explicit error.